File: 2241.cc
#include <iostream>
#include <stdio.h>
#include <string.h>
#include <algorithm>
using namespace std;
const int maxn = 10010;
struct Cube {
  int x, y, z;
  Cube(int a = 0, int b = 0, int c = 0) {
    x = a;
    y = b;
    z = c;
  }
  bool operator<(const Cube &t) const { return x * y > t.x * t.y; }
} cube[maxn];
int dp[maxn];
int main() {
  int n, tot, x, y, z, kase = 1;
  while (scanf("%d", &n), n) {
    for (int i = tot = 0; i < n; ++i) {
      scanf("%d%d%d", &x, &y, &z);
      cube[tot++] = Cube(x, y, z);
      cube[tot++] = Cube(y, z, x);
      cube[tot++] = Cube(x, z, y);
    }
    memset(dp, 0, sizeof dp);
    sort(cube, cube + tot);
    int maxHeight = 0;
    for (int i = 0; i < tot; ++i) {
      dp[i] = cube[i].z;
      for (int j = i - 1; j >= 0; --j) {
        if (cube[j].x > cube[i].x && cube[j].y > cube[i].y)
          dp[i] = max(dp[i], dp[j] + cube[i].z);
        if (cube[j].y > cube[i].x && cube[j].x > cube[i].y)
          dp[i] = max(dp[i], dp[j] + cube[i].z);
      }
      maxHeight = max(maxHeight, dp[i]);
    }
    printf("Case %d: maximum height = %d\n", kase++, maxHeight);
  }
  return 0;
}

File: 2241.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Scanner;

public class Main {

    /**
     * @param args
     */
    public static void main(String[] args) {

        Scanner cin = new Scanner(System.in);

        int kase = 0;
        while (true) {

            int n = cin.nextInt();

            if (n == 0) {
                break;
            }

            kase++;

            List list = new ArrayList();

            for (int i = 1; i <= n; i++) {
                int x = cin.nextInt();
                int y = cin.nextInt();
                int z = cin.nextInt();

                list.add(new T(x, y, z));
                list.add(new T(y, x, z));

                list.add(new T(x, z, y));
                list.add(new T(z, x, y));

                list.add(new T(y, z, x));
                list.add(new T(z, y, x));

            }

            Collections.sort(list, new Comparator() {

                public int compare(Object arg0, Object arg1) {
                    T t1 = (T) arg0;
                    T t2 = (T) arg1;

                    if (t2.x > t1.x) {
                        return t2.x - t1.x;
                    } else if (t2.x == t1.x && t2.y > t1.y) {
                        return t2.y - t1.y;
                    } else if (t2.x == t1.x && t2.y == t1.y && t2.z > t1.z) {
                        return t2.z - t1.z;
                    }

                    return -1;
                }
            });

            int[] f = new int[list.size() + 1];
            for (int i = 0; i < f.length; i++) {
                f[i] = 0;
            }

            f[0] = list.get(0).z;

            int max = f[0];

            for (int i = 1; i < list.size(); i++) {
                int m = list.get(i).z;

                for (int k = 0; k < i; k++) {

                    if (list.get(k).x > list.get(i).x && list.get(k).y > list.get(i).y) {
                        m = Math.max(m, f[k] + list.get(i).z);
                    }
                }

                max = Math.max(max, m);

                f[i] = m;
            }
            System.out.println("Case " + kase + ": maximum height = " + max);
        }
    }

    static class T {

        int x;
        int y;
        int z;

        public T(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }
    }

}

